File: ios/App/App/MindsPlugin.swift
//
//  MindsPlugin.swift
//  App
//
//  Created by Wennys on 26/04/23.
//

import Foundation
import Capacitor
import MindsSDK
import AVFAudio

@objc(MindsPlugin)
public class MindsPlugin: CAPPlugin, MindsSDKDelegate {
    
    public func onSuccess(_ response: BiometricResponse?) {
        self.biometricsReceive(response)
    }
    
    public func onError(_ response: BiometricResponse?) {
        self.biometricsReceive(response)
    }
    
    public func showMicrophonePermissionPrompt() {
        AVAudioSession.sharedInstance().requestRecordPermission { granted in
            print("granted: \(granted)")
        }
    }
    
    public func microphonePermissionNotGranted() {
        print("microphonePermissionNotGranted")
    }
    
    
    var sdk: MindsSDK?
    var navigationController: UINavigationController?
    var biometricsCall: CAPPluginCall?
    
    
    @objc func authentication(_ call: CAPPluginCall) {
        let document = call.getString("document") ?? ""
        let token = call.getString("token") ?? ""
        let telephone = call.getString("telephone") ?? ""
        
        biometricsCall = call
        self.startSDK(processType: .authentication, document: document, token: token, telephone: telephone, externalId: nil, externalCustomerId: nil)
    }
    
    @objc func enrollment(_ call: CAPPluginCall) {
        let document = call.getString("document") ?? ""
        let token = call.getString("token") ?? ""
        let telephone = call.getString("telephone") ?? ""
        
        biometricsCall = call
        self.startSDK(processType: .enrollment, document: document, token: token, telephone: telephone, externalId: nil, externalCustomerId: nil)
    }
    
    
    private func startSDK(processType: MindsSDK.ProcessType, document: String, token: String, telephone: String, externalId: String?, externalCustomerId: String?) {
        sdk = MindsSDK(delegate: self)
        sdk?.setToken(token)
        sdk?.setExternalId(externalId)
        sdk?.setExternalCustomerId(externalCustomerId)
        sdk?.setPhoneNumber(telephone)
        sdk?.setShowDetails(true)
        sdk?.setDocument(document)
        sdk?.setProcessType(processType)
        sdk?.setEnvironment(.staging)
        
        DispatchQueue.main.async {
            
            guard let navigationController: UINavigationController = self.bridge?.viewController?.navigationController else { return }
            
            
            self.sdk?.initialize(on: navigationController) { error in
                if let error = error {
                    do {
                        throw error
                    } catch DomainError.invalidDocument(let message) {
                        self.biometricsCall?.reject(message!, "invalid_document")
                        
                    } catch DomainError.invalidCPF(let message) {
                        self.biometricsCall?.reject(message!, "invalid_cpf")
                        
                    } catch DomainError.invalidPhoneNumber(let message) {
                        self.biometricsCall?.reject(message!, "invalid_phone_number")
                        
                    } catch DomainError.customerNotFoundToPerformVerification(let message) {
                        self.biometricsCall?.reject(message!, "customer_not_found")
                        
                    } catch DomainError.customerNotEnrolled(let message) {
                        self.biometricsCall?.reject(message!, "customer_not_enrolled")
                        
                    } catch DomainError.customerNotCertified(let message) {
                        self.biometricsCall?.reject(message!, "customer_not_certified")
                        
                    } catch DomainError.invalidToken {
                        self.biometricsCall?.reject("Invalid Token", "invalid_token")
                        
                    } catch DomainError.undefinedEnvironment {
                        self.biometricsCall?.reject("No environment defined", "undefined_environment")
                        
                    } catch DomainError.internalServerException {
                        self.biometricsCall?.reject("Internal server error", "internal_server_error")
                        
                    } catch {
                        print("\(error): \(error.localizedDescription)")
                        self.biometricsCall?.reject("ERROR", error.localizedDescription)
                    }
                }
            }
        }
    }
    
    
    private func biometricsReceive(_ response: BiometricResponse?) {
        self.biometricsCall?.resolve([
            "success": response?.success as Any,
            "error": [
                "code": response?.error?.code as Any,
                "description": response?.error?.description as Any
            ],
            "id": response?.id as Any,
            "cpf": response?.cpf as Any,
            "external_id": response?.externalID as Any,
            "created_at": response?.createdAt as Any,
            "utc_created_at": response?.utcCreatedAt as Any,
            "result": [
                "recommended_action": response?.result?.recommendedAction as Any,
                "reasons": response?.result?.reasons as Any
            ],
            "details": [
                "flag": [
                    "type": response?.details?.flag?.type as Any,
                    "status": response?.details?.flag?.status as Any
                ],
                "liveness": [
                    "status": response?.details?.liveness?.status as Any,
                    "replay_attack": [
                        "enabled": response?.details?.liveness?.replayAttack?.enabled as Any,
                        "status": response?.details?.liveness?.replayAttack?.status as Any,
                        "result": response?.details?.liveness?.replayAttack?.result as Any,
                        "confidence": response?.details?.liveness?.replayAttack?.confidence as Any,
                        "score": response?.details?.liveness?.replayAttack?.score as Any,
                        "threshold": response?.details?.liveness?.replayAttack?.threshold as Any
                    ],
                    "deepfake": [
                        "enabled": response?.details?.liveness?.deepFake?.enabled as Any,
                        "status": response?.details?.liveness?.deepFake?.status as Any,
                        "result": response?.details?.liveness?.deepFake?.result as Any,
                        "confidence": response?.details?.liveness?.deepFake?.confidence as Any,
                        "score": response?.details?.liveness?.deepFake?.score as Any,
                        "threshold": response?.details?.liveness?.deepFake?.threshold as Any
                    ],
                    "sentence_match": [
                        "enabled": response?.details?.liveness?.sentenceMatch?.enabled as Any,
                        "status": response?.details?.liveness?.sentenceMatch?.status as Any,
                        "result": response?.details?.liveness?.sentenceMatch?.result as Any,
                        "confidence": response?.details?.liveness?.sentenceMatch?.confidence as Any,
                        "score": response?.details?.liveness?.sentenceMatch?.score as Any,
                        "threshold": response?.details?.liveness?.sentenceMatch?.threshold as Any
                    ]
                ],
                "voice_match": [
                    "status": response?.details?.voiceMatch?.status as Any,
                    "result": response?.details?.voiceMatch?.result as Any,
                    "confidence": response?.details?.voiceMatch?.confidence as Any,
                    "score": response?.details?.voiceMatch?.score as Any,
                    "threshold": response?.details?.voiceMatch?.threshold as Any
                ]
            ]
            
        ])
    }
}

File: android/app/src/main/java/io/ionic/starter/MindsPlugin.kt
package io.ionic.starter
import androidx.activity.result.ActivityResult
import com.getcapacitor.*
import com.getcapacitor.annotation.ActivityCallback
import com.getcapacitor.annotation.CapacitorPlugin
import digital.minds.clients.sdk.android.MindsDigital
import digital.minds.clients.sdk.kotlin.data.model.VoiceMatchResponse
import digital.minds.clients.sdk.kotlin.domain.constants.VOICE_MATCH_RESPONSE
import digital.minds.clients.sdk.kotlin.domain.exceptions.*
import digital.minds.clients.sdk.kotlin.main.MindsSDK
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import org.json.JSONArray
import org.json.JSONObject


@CapacitorPlugin(name = "Minds")
class MindsPlugin : Plugin() {

  private lateinit var authenticationMindsSDK: MindsSDK
  private lateinit var enrollmentMindsSDK: MindsSDK
  private lateinit var _result: PluginCall


  @PluginMethod
  fun authentication(call: PluginCall) {
    val document: String? = call.getString("document")
    val token: String? = call.getString("token")
    val telephone: String? = call.getString("telephone")
    _result = call

    try {
      authenticationMindsSDK =
        MindsConfig.authentication(document!!, token!!, telephone!!)
      CoroutineScope(Dispatchers.Main).launch {
        try {
          val intent = MindsDigital.getIntent(context, authenticationMindsSDK)
          startActivityForResult(call, intent, "result")
        } catch (e: InvalidDocument) {
          _result.reject(e.message, "invalid_document")
        } catch (e: InvalidCPF) {
          _result.reject(e.message, "invalid_cpf")
        } catch (e: InvalidPhoneNumber) {
          _result.reject(e.message, "invalid_phone_number")
        } catch (e: CustomerNotFoundToPerformVerification) {
          _result.reject(e.message, "customer_not_found")
        } catch (e: CustomerNotEnrolled) {
          _result.reject(e.message, "customer_not_enrolled")
        } catch (e: CustomerNotCertified) {
          _result.reject(e.message, "customer_not_certified")
        } catch (e: InvalidToken) {
          _result.reject(e.message, "invalid_token")
        } catch (e: InternalServerException) {
          _result.reject(e.message, "internal_server_error")
        } catch (e: Exception) {
          _result.reject(e.message, "MINDS_SDK_INIT_ERROR")
        }
      }
    } catch (e: Exception) {
      _result.reject(e.message, "MINDS_SDK_INIT_ERROR")
    }
  }

  @PluginMethod
  fun enrollment(call: PluginCall) {
    val document: String? = call.getString("document")
    val token: String? = call.getString("token")
    val telephone: String? = call.getString("telephone")
    _result = call

    try {
      enrollmentMindsSDK =
        MindsConfig.enrollment(document!!, token!!, telephone!!)
      CoroutineScope(Dispatchers.Main).launch {
        try {
          val intent = MindsDigital.getIntent(context, enrollmentMindsSDK)
          startActivityForResult(call, intent, "result")
        } catch (e: InvalidDocument) {
          _result.reject(e.message, "invalid_document")
        } catch (e: InvalidCPF) {
          _result.reject(e.message, "invalid_cpf")
        } catch (e: InvalidPhoneNumber) {
          _result.reject(e.message, "invalid_phone_number")
        } catch (e: CustomerNotFoundToPerformVerification) {
          _result.reject(e.message, "customer_not_found")
        } catch (e: CustomerNotEnrolled) {
          _result.reject(e.message, "customer_not_enrolled")
        } catch (e: CustomerNotCertified) {
          _result.reject(e.message, "customer_not_certified")
        } catch (e: InvalidToken) {
          _result.reject(e.message, "invalid_token")
        } catch (e: InternalServerException) {
          _result.reject(e.message, "internal_server_error")
        } catch (e: Exception) {
          _result.reject(e.message, "MINDS_SDK_INIT_ERROR")
        }
      }
    } catch (e: Exception) {
      _result.reject(e.message, "MINDS_SDK_INIT_ERROR")
    }
  }

  @ActivityCallback
  private fun result(call: PluginCall?, result: ActivityResult) {
    if (call == null) {
      return
    }
    val mindsSDKResponse = result.data?.extras?.get(VOICE_MATCH_RESPONSE) as? VoiceMatchResponse
            val jsonObject = JSONObject()
            jsonObject.put("success", mindsSDKResponse?.success)
            jsonObject.put("error", JSONObject().apply {
                put("code", mindsSDKResponse?.error?.code)
                put("description", mindsSDKResponse?.error?.description)
            })
            jsonObject.put("id", mindsSDKResponse?.id)
            jsonObject.put("cpf", mindsSDKResponse?.cpf)
            jsonObject.put("external_id", mindsSDKResponse?.externalId)
            jsonObject.put("created_at", mindsSDKResponse?.createdAt)
            jsonObject.put("utc_created_at", mindsSDKResponse?.utcCreatedAt)
            jsonObject.put("result", JSONObject().apply {
                put("recommended_action", mindsSDKResponse?.result?.recommendedAction)
                put("reasons", JSONArray(mindsSDKResponse?.result?.reasons))
            })
            jsonObject.put("details", JSONObject().apply {
                jsonObject.put("flag", JSONObject().apply {
                    put("type", mindsSDKResponse?.details?.flag?.type)
                    put("status", mindsSDKResponse?.details?.flag?.status)
                })
                put("liveness", JSONObject().apply {
                    put("status", mindsSDKResponse?.details?.liveness?.status)
                    put("replay_attack", JSONObject().apply {
                        put("enabled", mindsSDKResponse?.details?.liveness?.replayAttack?.enabled)
                        put("status", mindsSDKResponse?.details?.liveness?.replayAttack?.status)
                        put("result", mindsSDKResponse?.details?.liveness?.replayAttack?.result)
                        put("confidence", mindsSDKResponse?.details?.liveness?.replayAttack?.confidence)
                        put("score", mindsSDKResponse?.details?.liveness?.replayAttack?.score)
                        put("threshold", mindsSDKResponse?.details?.liveness?.replayAttack?.threshold)
                    })
                    put("deepfake", JSONObject().apply {
                        put("enabled", mindsSDKResponse?.details?.liveness?.deepFake?.enabled)
                        put("status", mindsSDKResponse?.details?.liveness?.deepFake?.status)
                        put("result", mindsSDKResponse?.details?.liveness?.deepFake?.result)
                        put("confidence", mindsSDKResponse?.details?.liveness?.deepFake?.confidence)
                        put("score", mindsSDKResponse?.details?.liveness?.deepFake?.score)
                        put("threshold", mindsSDKResponse?.details?.liveness?.deepFake?.threshold)
                    })
                    put("sentence_match", JSONObject().apply {
                        put("enabled", mindsSDKResponse?.details?.liveness?.sentenceMatch?.enabled)
                        put("status", mindsSDKResponse?.details?.liveness?.sentenceMatch?.status)
                        put("result", mindsSDKResponse?.details?.liveness?.sentenceMatch?.result)
                        put("confidence", mindsSDKResponse?.details?.liveness?.sentenceMatch?.confidence)
                        put("score", mindsSDKResponse?.details?.liveness?.sentenceMatch?.score)
                        put("threshold", mindsSDKResponse?.details?.liveness?.sentenceMatch?.threshold)
                    })
                })
                put("voice_match", JSONObject().apply {
                    put("status", mindsSDKResponse?.details?.voiceMatch?.status)
                    put("result", mindsSDKResponse?.details?.voiceMatch?.result)
                    put("confidence", mindsSDKResponse?.details?.voiceMatch?.confidence)
                    put("score", mindsSDKResponse?.details?.voiceMatch?.score)
                    put("threshold", mindsSDKResponse?.details?.voiceMatch?.threshold)
                })
            })
    val jsonResult = JSObject.fromJSONObject(jsonObject)
    _result.resolve(jsonResult)
  }
}

